### PR TITLE
fix(devtools): Fix pm2 config runs on Windows

### PR DESCRIPTION
**Pull request by chentong7**
### description
Each of those packages does include the listed script, but something about how PM2 is configured or operating it's not putting the commands together properly. PM2 needs cmd.exe as interpreter in order to drive the node.js application.

### sample

[AB#5471](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/5471)

### CODE REVIEW

**Comment by alexvy86** [comment]
@chentong7 I tried this and it doesn't work... pm2 says it started the processes and they're online, but nothing is actually running. The logs only seem to show that a `cmd` was started for each of the 4 apps in the config file, but the npm scripts do not run.

@rohandubal @Josmithr @seanimam I'm pretty convinced pm2 is just not maintained enough to fix bugs that only impact Windows, so I think at this point we should explicitly drop support for cross-package hot reloading on Windows and document how to do it manually in devtools-example's README. I'm even tempted to drop it entirely and just stick to the manual approach, so we don't have differing workflows per OS. Thoughts?

**Comment by Josmithr** [comment]
> @chentong7 I tried this and it doesn't work... pm2 says it started the processes and they're online, but nothing is actually running. The logs only seem to show that a `cmd` was started for each of the 4 apps in the config file, but the npm scripts do not run.
> 
> @rohandubal @Josmithr @seanimam I'm pretty convinced pm2 is just not maintained enough to fix bugs that only impact Windows, so I think at this point we should explicitly drop support for cross-package hot reloading on Windows and document how to do it manually in devtools-example's README. I'm even tempted to drop it entirely and just stick to the manual approach, so we don't have differing workflows per OS. Thoughts?

Yeah, I think that's fine. That's how all our other sample apps work. If this was easy to maintain, it would be nice to keep. But if it's not, documenting the manual steps should be fine IMO.

**Comment by tylerbutler** [comment]
> I'm pretty convinced pm2 is just not maintained enough to fix bugs that only impact Windows, so I think at this point we should explicitly drop support for cross-package hot reloading on Windows and document how to do it manually in devtools-example's README. I'm even tempted to drop it entirely and just stick to the manual approach, so we don't have differing workflows per OS. Thoughts?

[nodemon ](https://github.com/remy/nodemon/) is a well-maintained alternative depending on what you're using pm2 for.

**Comment by chentong7** [comment]
> > @chentong7 I tried this and it doesn't work... pm2 says it started the processes and they're online, but nothing is actually running. The logs only seem to show that a `cmd` was started for each of the 4 apps in the config file, but the npm scripts do not run.
> > @rohandubal @Josmithr @seanimam I'm pretty convinced pm2 is just not maintained enough to fix bugs that only impact Windows, so I think at this point we should explicitly drop support for cross-package hot reloading on Windows and document how to do it manually in devtools-example's README. I'm even tempted to drop it entirely and just stick to the manual approach, so we don't have differing workflows per OS. Thoughts?
> 
> Yeah, I think that's fine. That's how all our other sample apps work. If this was easy to maintain, it would be nice to keep. But if it's not, documenting the manual steps should be fine IMO.

Modified accordingly. Now it will run the script based on os

**Comment by alexvy86** [comment]
> [nodemon ](https://github.com/remy/nodemon/) is a well-maintained alternative depending on what you're using pm2 for.

I did look into it but didn't seem like it supports managing multiple processes. Thanks for the suggestion though :).

**Review by alexvy86 — COMMENTED** [review]
Ok, good news! The addition of `concurrently` to the deps made me wonder if we could make do simply with that and I found a good solution.

With these npm script definitions, things seem to work as expected in both Linux and Windows:

```json
{
	"start": "concurrently \"npm:watch:*\" \"npm:start:client\"",
	"watch:devtools": "tsc --watch --preserveWatchOutput --project ../devtools/tsconfig.json",
	"watch:devtools-core": "tsc --watch --preserveWatchOutput --project ../devtools-core/tsconfig.json",
	"watch:devtools-view": "tsc --watch --preserveWatchOutput --project ../devtools-view/tsconfig.json",
}
```
Saves us the `cd` from the current approach to run `tsc` on each project. And added `--preserveWatchOutput` so that it doesn't clear the console when it starts a rebuild, because that clears the logs from all the processes we're running with `concurrently`.

Can you try that locally @chentong7 to make sure it works for someone else? We can get rid of the `watch` script, and completely remove pm2 (the ecosystem file and the dev dependency).

Also I think there was a bad merge resolution when you brought in main, that kept the older versions of several `eslint*` packages. We want to keep those at whatever version main says they should be.

**Comment by Josmithr** [comment]
> Ok, good news! The addition of `concurrently` to the deps made me wonder if we could make do simply with that and I found a good solution.
> 
> With these npm script definitions, things seem to work as expected in both Linux and Windows:
> 
> ```json
> {
> 	"start": "concurrently \"npm:watch:*\" \"npm:start:client\"",
> 	"watch:devtools": "tsc --watch --preserveWatchOutput --project ../devtools/tsconfig.json",
> 	"watch:devtools-core": "tsc --watch --preserveWatchOutput --project ../devtools-core/tsconfig.json",
> 	"watch:devtools-view": "tsc --watch --preserveWatchOutput --project ../devtools-view/tsconfig.json",
> }
> ```
> 
> Saves us the `cd` from the current approach to run `tsc` on each project. And added `--preserveWatchOutput` so that it doesn't clear the console when it starts a rebuild, because that clears the logs from all the processes we're running with `concurrently`.
> 
> Can you try that locally @chentong7 to make sure it works for someone else? We can get rid of the `watch` script, and completely remove pm2 (the ecosystem file and the dev dependency).
> 
> Also I think there was a bad merge resolution when you brought in main, that kept the older versions of several `eslint*` packages. We want to keep those at whatever version main says they should be.

I like this path forward. It's simple and uses a library we already use heavily across the repo (`concurrently`). +1 from Josh 🙂

**Comment by seanimam** [comment]
I am liking the use of concurrently if it simplifies things and works cross platform!

**Comment by alexvy86** [comment]
@chentong7 for the policy bits failing right now, we'll need to add the names of the new scripts that call `tsc` to [this list](https://github.com/microsoft/FluidFramework/blob/b3a948a059215fe843103874249bc7895eabdbd9/fluidBuild.config.cjs#L305). One of our policies is trying to ensure that if a script calls `tsc` somehow, the build graph must ensure a dependency on the `tsc` task, but that's for a different use case than what we're doing here.

The test failures seem transient, they might not happen when you add the commit to exclude the new script from that policy.

**Comment by alexvy86** [comment]
@chentong7 I also noticed the lockfile seems to have a lot of changes that shouldn't be there; it seems like changes that it had in main were dropped during a merge. I suggest restoring that file to its current state in main, then running `pnpm i` again from this branch so only the few necessary changes show up.

**Comment by chentong7** [comment]
> @chentong7 for the policy bits failing right now, we'll need to add the names of the new scripts that call `tsc` to [this list](https://github.com/microsoft/FluidFramework/blob/b3a948a059215fe843103874249bc7895eabdbd9/fluidBuild.config.cjs#L305). One of our policies is trying to ensure that if a script calls `tsc` somehow, the build graph must ensure a dependency on the `tsc` task, but that's for a different use case than what we're doing here.
> 
> The test failures seem transient, they might not happen when you add the commit to exclude the new script from that policy.

Thx!!!!!

**Review by alexvy86 — APPROVED** [review]
🚀

**Review by alexvy86 — CHANGES_REQUESTED** [review]
Actually, holding the approve for a sec, we might be able to do the exclusions for policy in a more targeted manner, looking into it

File: fluidBuild.config.cjs [hunk 1]
```diff
@@ -302,7 +302,12 @@ module.exports = {
 		],
 		fluidBuildTasks: {
 			tsc: {
-				ignoreTasks: ["tsc:watch"],
+				ignoreTasks: [
+					"tsc:watch",
+					"watch:devtools",
+					"watch:devtools-core",
+					"watch:devtools-view",
+				],
```
[the remark this answers is not in the document]
Comment (alexvy86) [reply fluidBuild.config.cjs:310]: @chentong7 can you try this ⬆️ ? You can run the policy check locally so we don't have to wait for pipeline runs to complain (or not). If we can put the new exclusions in devtools-example's `package.json` it'd be much better.

[the remark this answers is not in the document]
Comment (chentong7) [reply fluidBuild.config.cjs:310]: On it!

[the remark this answers is not in the document]
Comment (chentong7) [reply fluidBuild.config.cjs:310]: seems like fluidBuild.config.cjs is the only way :clown_face:

[the remark this answers is not in the document]
Comment (tylerbutler) [reply fluidBuild.config.cjs:310]: I would prefer to add a handler exclusion to policy.handlerExclusions, like so:

```ts
			"fluid-build-tasks-tsc": [
				"packages/tools/devtools/devtools-core/package.json",
				"packages/tools/devtools/devtools-example/package.json",
				"packages/tools/devtools/devtools-view/package.json",
			],
```

#18102 will make it easier to know what policy is failing so it's easier to exclude if needed.

[the remark this answers is not in the document]
Comment (tylerbutler) [reply fluidBuild.config.cjs:310]: And #18104 should remove the need for the exclusion.

[the remark this answers is not in the document]
Comment (alexvy86) [reply fluidBuild.config.cjs:310]: @chentong7 now that Tyler's PR is merged, let's just undo adding the new scripts to the exclusion list, and this should be good to go.

[the remark this answers is not in the document]
Comment (chentong7) [reply fluidBuild.config.cjs:310]: okok

**Review by alexvy86 — APPROVED** [review]

**Comment by alexvy86** [comment]
Waiting for the client release group to consume a version of build-tools that includes #18104 .

**Comment by tylerbutler** [comment]
> Waiting for the client release group to consume a version of build-tools that includes #18104 .

https://github.com/microsoft/FluidFramework/pull/18263 should have it.

File: packages/tools/devtools/devtools-example/package.json [hunk 2]
```diff
@@ -26,14 +26,18 @@
 		"prettier": "prettier --check . --ignore-path ../../../../.prettierignore",
 		"prettier:fix": "prettier --write . --ignore-path ../../../../.prettierignore",
 		"rebuild": "npm run clean && npm run build",
-		"start": "pm2 start pm2.ecosystem.config.js",
+		"start": "concurrently \"npm:watch:*\" \"npm:start:client\"",
 		"start:client": "webpack serve --config webpack.config.js",
 		"stop": "pm2 stop pm2.ecosystem.config.js",
```
Comment (seanimam) [inline packages/tools/devtools/devtools-example/package.json:31]: Just to confirm, do we want to  keep remaining pm2 script related stuff around?
There is readme documentation and the entire pm2.ecosystem.config.js file

**Review by seanimam — DISMISSED** [review]

File: packages/tools/devtools/devtools-example/package.json (hunk 2 above)
Comment (alexvy86) [reply packages/tools/devtools/devtools-example/package.json:31]: Good catch, we should remove the `stop` script and the references in the docs/readme. @chentong7 can you do that when you have a chance?

Comment (chentong7) [reply packages/tools/devtools/devtools-example/package.json:31]: Gotcha

**Review by alexvy86 — COMMENTED** [review]
@chentong7 I see that for some reason we can't remove the new npm scripts from the policy exclusions or they keep complaining. We can chat with @tylerbutler about that once he's back. But we still need to remove the pm2 ecosystem file and all references to pm2 from the readme. Can you open another PR for that?

**Pull request merged**